Block Bindings: Disable editing of bound block attributes in editor UI

SantosGuillamot opened the pull request:
## What?
As discussed in this Slack [thread](https://wordpress.slack.com/archives/C02QB2JS7/p1705917394175109), I'm splitting [this other pull request](https://github.com/WordPress/gutenberg/pull/58059) to handle first just the use case where the editor understands the bindings attribute and adapt the UI accordingly.

As part of that, in this PR I'm doing:
* Provide an action to register a new source in the editor: [link](https://github.com/WordPress/gutenberg/compare/add/block-bindings-support-in-the-editor?expand=1#diff-4d145a0c43689f5a56c00a8e2919550768ebd233e91a6c6d04fee71a44bae9f2R364-R372).
* Provide some selectors to get all the registered sources or a specific one: [link](https://github.com/WordPress/gutenberg/compare/add/block-bindings-support-in-the-editor?expand=1#diff-23e851c3077ed95d23bb1669da8bafbae3f1e2e952897987e45b49555004dd83R289-R296).
* Add filter to the `BlockEdit` to modify `attributes` and `setAttributes` and make it understand the bindings: [link](https://github.com/WordPress/gutenberg/compare/add/block-bindings-support-in-the-editor?expand=1#diff-99d79bdba99fe4517413b0fd9503eef924e439c03d44504c1ad0606b727cebb6R31-R100).
* Use addFilter to add the needed context to all the allowed blocks: [link](https://github.com/WordPress/gutenberg/compare/add/block-bindings-support-in-the-editor?expand=1#diff-99d79bdba99fe4517413b0fd9503eef924e439c03d44504c1ad0606b727cebb6R126-R144).
* Adapt the paragraph, heading, button, and image blocks UI to identify when bindings exist.
* Register the post meta source using these mechanisms: [link](https://github.com/WordPress/gutenberg/compare/add/block-bindings-support-in-the-editor?expand=1#diff-cf4447b4134dbf24fa2819703999f0131449fb79a0611bd544c5e05fd80aa36bR21-R47).

## Why?
Splitting the original pull request should make it easier to review and it could help us land something sooner that covers understanding the bindings from the editor, although it doesn't include any UI to create/modify the bindings.

## Testing Instructions

For all the testing, we have to go to Gutenberg-Experiments and enable the Test Block Bindings option.

0. Register a new custom field however you prefer. You can use a snippet similar to this:

```PHP
register_meta(
	'post',
	'text_custom_field',
	array(
		'show_in_rest' => true,
		'single'       => true,
		'type'         => 'string',
		'default'	   => 'This is the content of the text custom field',
	)
);
register_meta(
	'post',
	'url_custom_field',
	array(
		'show_in_rest' => true,
		'single'       => true,
		'type'         => 'string',
		'default'	   => 'https://wpmovies.dev/wp-content/uploads/2023/04/goncharov-poster-original-1-682x1024.jpeg',
	)
);
```

### In a page

**_Test paragraph_**
1. Add a paragraph with the content connected to a custom field:

```
<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":{"name":"post_meta","attributes":{"value":"text_custom_field"}}}}}} -->
<p>Content of the PAGE text custom field</p>
<!-- /wp:paragraph -->
```

2. Add a paragraph without any bindings.
3. Check that the paragraph with the bindings became non-editable.
4. Checkthat the paragraph shows the content of the custom field.
5. Check that the normal paragraph works as expected.
6. Go to the frontend and check that it works as well there.

**_Test heading_**

Repeat the paragraph test but using a heading.

**_Test button_**

1. Add a button with the text connected to a custom field.
2. Add another button with the URL connected to a custom field.
3. Add a button with the text and the URL connected to custom fields.

```
<!-- wp:buttons -->
<div class="wp-block-buttons"><!-- wp:button {"metadata":{"bindings":{"text":{"source":{"name":"post_meta","attributes":{"value":"text_custom_field"}}}}}} -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button" href="https://wordpress.org">ANY TEXT</a></div>
<!-- /wp:button -->

<!-- wp:button {"metadata":{"bindings":{"url":{"source":{"name":"post_meta","attributes":{"value":"url_custom_field"}}}}}} -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button" href="https://wordpress.org">URL</a></div>
<!-- /wp:button -->

<!-- wp:button {"metadata":{"bindings":{"text":{"source":{"name":"post_meta","attributes":{"value":"text_custom_field"}}},"url":{"source":{"name":"post_meta","attributes":{"value":"url_custom_field"}}}}}} -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button" href="https://wordpress.org">BOTH</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->
```

4. Check that for the buttons with the text connected, they are not editable and they show the content of the custom field.
5. Check that for the buttons with the URL connected, the buttons to change the URL disappear from the UI.
6. Check that everything works in the frontend.

**_Test image_**

1. Add an image with the URL connected to a custom field.
2. Add an image with the alt attribute connected to a custom field.
3. Add an image with the title attribute connected to a custom field.
4. Add an image with more than one connection.

```
<!-- wp:heading {"level":3} -->
<h3 class="wp-block-heading">Image url and alt connected</h3>
<!-- /wp:heading -->

<!-- wp:image {"id":135,"sizeSlug":"large","linkDestination":"none","metadata":{"bindings":{"url":{"source":{"name":"post_meta","attributes":{"value":"url_custom_field"}}},"alt":{"source":{"name":"post_meta","attributes":{"value":"text_custom_field"}}}}}} -->
<figure class="wp-block-image size-large"><img src="https://wpmovies.dev/wp-content/uploads/2023/03/3bhkrj58Vtu7enYsRolD1fZdja1-683x1024.jpg" alt="Content of the PAGE text custom field" class="wp-image-135" title="Image title"/></figure>
<!-- /wp:image -->

<!-- wp:heading {"level":3} -->
<h3 class="wp-block-heading">Alt connected</h3>
<!-- /wp:heading -->

<!-- wp:image {"id":133,"sizeSlug":"large","linkDestination":"none","metadata":{"bindings":{"alt":{"source":{"name":"post_meta","attributes":{"value":"text_custom_field"}}}}}} -->
<figure class="wp-block-image size-large"><img src="http://custom-fields.local/wp-content/uploads/2023/11/1-1024x1024.jpg" alt="Content of the PAGE text custom field" class="wp-image-133" title="Image title"/></figure>
<!-- /wp:image -->
```

5. For the images with the URL connected, check that the related buttons to link an image have disappeared and the image shows the URL from the custom field.
6. For the images with the alt/title connected, check that, in the right sidebar, they are disabled. Showing the value of the custom field instead and a message saying it is connected to custom fields.

### In a template

Go to a page template, for example, and repeat the process. In this case, the blocks can't show the value of the custom fields because it depends on each page. They should show a placeholder instead.

## Remaining issues
I'll try to gather here the remaining issues related to this pull request. So far I'd like to work on:
* Fix the query loop not showing the content properly.
* Review the whole UX.

## Screenshots or screencast

https://github.com/WordPress/gutenberg/assets/34552881/dccbf5f1-2148-4297-8a63-f7dc6fb5e043

SantosGuillamot committed:
Add actions and selectors to register new sources

SantosGuillamot committed:
Add hook to read the bindings attribute in Edit

SantosGuillamot committed:
Add context to all the blocks with bindings

SantosGuillamot committed:
Lock rich text when `isContentBound` is true

SantosGuillamot committed:
Adapt paragraph and heading blocks UI

SantosGuillamot committed:
Adapt button block UI

SantosGuillamot committed:
Adapt image block UI

SantosGuillamot committed:
Register post meta source

michalczaplinski commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -376,7 +376,7 @@ export function RichTextWrapper(
 					anchorRef,
 				] ) }
-				contentEditable={ true }
+				contentEditable={ props.isContentBound ? false : true }
```
I'm not sure if it's me but I found the name confusing. "contentBound" sounds like something is "bound for" (destined for, enroute to) the content, not that we are binding **to** to the content.

SantosGuillamot replied:
I've changed that part in [this commit](https://github.com/WordPress/gutenberg/pull/58085/commits/a8a6da3508405cf0772b132ad1ca204b2bc6da1a#diff-bec072c72a2f961596701aef67f67575b0a3b918135a6587229a0cd79bb3db44R404). It now uses `shouldDisableEditing`. Is that clearer?

michalczaplinski replied:
yes, perfect! Sorry, I didn't notice the change.

artemiomorales reviewed (commented):
EDIT: Turns out this was an input error on my end, please disregard.

I've taken an initial look and it appears that it works well in the Post Editor, but in the Site Editor I've defined the following layout, including a bound image and some buttons: 

![Screenshot 2024-01-22 at 5 32 49 PM](https://github.com/WordPress/gutenberg/assets/5360536/bc382188-63a2-40b2-9775-3d1cdf83a9e7)

However, the image and two of the buttons disappear when viewing the site:

![Screenshot 2024-01-22 at 5 33 02 PM](https://github.com/WordPress/gutenberg/assets/5360536/43468bb1-e116-4ae2-90c2-0dd1a6e75099)

Anyone else seeing this?
I'll continue looking at the code and aim to offer additional insights tomorrow.

On the code side, I'm still working through it but initial impressions are good; I don't have extensive experience with Gutenberg though and think it'd be good to get a review by someone who can speak to potential pitfalls and offer additional insight.

artemiomorales commented on packages/block-editor/src/store/private-actions.js:
```diff
@@ -360,3 +360,13 @@ export function stopEditingAsBlocks( clientId ) {
+		sourceComponent: source.component,
+		useSource: source.useSource,
+	};
+}
```
Since this is private API, we don't need to worry about this `registerBlockBindingsSource()` being perfect or backwards compatible, right?

SantosGuillamot replied:
That's it. Keeping it private, and only used in Gutenberg, makes it easier to change it in the future without worrying too much.

SaxonF commented:
Looking good, thanks @SantosGuillamot !

A couple suggestions:

- When in a template, for paragraph and heading fields connected to a field, let's not dim the cololur at all. It should be accurate to whatever styling is applied to the block
- When in a template, for images, we could still show the alternative text field except it would be disabled and with the default value of the field inserted.

I don't see any design blockers for 6.5 here. During beta testing we might pick up the need to better highlight blocks that are connected to a field in some way.

gziolo commented:
There are e2e test failures logged for Pattern Overrides so it needs to be investigated. I suspect, it’s no longer possible to edit content for the blocks that have overrides enabled in the pattern.

gziolo commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -376,7 +376,7 @@ export function RichTextWrapper(
 					anchorRef,
 				] ) }
-				contentEditable={ true }
+				contentEditable={ props.isContentBound ? false : true }
```
@ellatrix, what alternatives are there to put `RichText` in read-only mode that don't require introducing a new prop?

I'm also wondering what should happen for all these users using assistive technology. It feels like we should ensure it still possible to move focus when tabbing to the HTML element as in some cases RichText is the block wrapper so it could make it harder navigating to blocks like Paragraph or Heading.

gziolo replied:
Noting also that this new prop fails one of the unit tests:

<img width="1500" alt="Screenshot 2024-01-23 at 10 03 46" src="https://github.com/WordPress/gutenberg/assets/699132/6fa61a7f-dee7-4529-9ae1-f7fb21cd92b5">

SantosGuillamot replied:
I believe we can get directly the block attributes from the `RichText` component and read the `metadata.bindings` there. I'll give it a try.

SantosGuillamot replied:
I made [this commit](https://github.com/WordPress/gutenberg/pull/58085/commits/a8a6da3508405cf0772b132ad1ca204b2bc6da1a) to reflect what I was referring to. It seems cleaner to me, but I am not familiar enough with RichText to say if that's the right approach.

michalczaplinski replied:
Also noting that for me, selecting a Paragraph block while in the "edit" mode
requires a double click. In the "select" mode it seems to work fine.

michalczaplinski replied:
I also notice that selecting the Paragraph block that is connected to a source requires a double click in the "edit" mode. In the "select" mode it seems that it works just fine.


https://github.com/WordPress/gutenberg/assets/5417266/64f77fef-660c-46ea-8870-2ac10f4ba407

SantosGuillamot replied:
I agree it's not the best UX. I think that's the way the RichText works, right now when the editing is locked, so I am not sure if that's on purpose or we should look for a different approach.

SantosGuillamot replied:
This should be fixed with [this commit](https://github.com/WordPress/gutenberg/pull/58085/commits/42462609211628115cd9b03cb4758228a87ab103).

gziolo commented on packages/block-library/src/button/edit.js:
```diff
@@ -296,7 +298,7 @@ function ButtonEdit( props ) {
 					/>
 				) }
-				{ isURLSet && isLinkTag && (
+				{ isURLSet && isLinkTag && ! metadata?.bindings?.url && (
```
What if someone would like to unlink the button? In general, should it be possible to remove the binding in the UI?

SantosGuillamot replied:
> In general, should it be possible to remove the binding in the UI?

I wasn't sure about that one. My reasoning was that, as right now bindings can only be added through the code editor, it shouldn't be a blocker if they can only be removed through the code editor as well. When we improve the UI and add the possibility to create bindings, we should provide mechanisms to remove them as well.

Anyway, I don't know which is the correct approach. Any thoughts?

gziolo commented on packages/editor/src/bindings/post-meta.js:
```diff
@@ -0,0 +1,47 @@
+
+// Prettify the name until the label is available in the REST API endpoint.
+const keyToLabel = ( key ) => {
+	return key
```
It isn’t the most user friendly to show in the UI machine readable key that can’t be localized. Maybe it’s fine for someone technical creating the site, but far from perfect for someone editing the content.

SantosGuillamot replied:
I agree. I saw that as a workaround. The issue is that we don't have labels in the meta fields yet. Do you think it would be better to just pass the label? `text_custom_field` for example.

gziolo replied:
Maybe that would be more appropriate as it would be clear it’s a code name. At least until we support labels. @SaxonF, what do you think?

SantosGuillamot replied:
I was considering using something like "Connected to { meta_key } from { source }". Which would translate into "Connected to my_custom_field from Post Meta". Maybe it is too long for some cases like a button text.

gziolo commented on packages/editor/src/bindings/post-meta.js:
```diff
@@ -0,0 +1,47 @@
+
+export default {
+	name: 'post_meta',
+	label: 'Post Meta',
```
It needs to be translatable.

SantosGuillamot replied:
Is using `__( 'Post Meta' )` enough?

gziolo replied:
Yes.

SantosGuillamot replied:
Done in [this commit](https://github.com/WordPress/gutenberg/pull/58085/commits/96449467ec05e18806d6bf4dadb92a16feba0ec5).

gziolo commented on packages/editor/src/bindings/post-meta.js:
```diff
@@ -0,0 +1,47 @@
+export default {
+	name: 'post_meta',
+	label: 'Post Meta',
+	component: null,
```
What’s the role of this setting? It doesn’t seem to be needed at the moment.

SantosGuillamot replied:
It will potentially be used in the future when we have the UI to create the bindings. Should I remove it until then?

gziolo replied:
Yes, let’s keep only usable code.

SantosGuillamot replied:
Done in [this commit](https://github.com/WordPress/gutenberg/pull/58085/commits/1984749af6502fe017903c15f6f5f6b05c21ae8c).

gziolo commented on packages/editor/src/bindings/post-meta.js:
```diff
@@ -0,0 +1,47 @@
+ */
+import { store as editorStore } from '../store';
+
+const { getCurrentPostType } = select( editorStore );
```
It feels like this should be moved to `useSource` to ensure it always triggers rerenders whenever the post type changes.

SantosGuillamot replied:
Okay, I'll move it there 🙂

SantosGuillamot replied:
Done in [this commit](https://github.com/WordPress/gutenberg/pull/58085/commits/478f861c2f7a5ca360617ee9bb40b7aa7f87d22d).

gziolo commented on packages/editor/src/bindings/post-meta.js:
```diff
@@ -0,0 +1,47 @@
+		if ( postType === 'wp_template' ) {
+			return { placeholder: keyToLabel( metaKey ) };
+		}
+		const metaValue = meta[ metaKey ];
```
What should happen when the `metaValue` is empty?

SantosGuillamot replied:
Maybe we should always pass a `placeholder` and change the logic [here](https://github.com/WordPress/gutenberg/pull/58085/files/a4dc34a9eaa717c953b0393c82398ec28e9c62c7#diff-99d79bdba99fe4517413b0fd9503eef924e439c03d44504c1ad0606b727cebb6R69-R76) to check if the value is empty. And in that case, use the placeholder instead of the value?

gziolo replied:
That might work better if we always have a fallback available and here we would only have to read the value.

SantosGuillamot replied:
Done in [this commit](https://github.com/WordPress/gutenberg/pull/58085/commits/35155385e8fd4099c9f5e48b5042e5f3880ce7e1).

SantosGuillamot commented:
> I've taken an initial look and it appears that it works well in the Post Editor, but in the Site Editor I've defined the following layout, including a bound image and some buttons. However, the image and two of the buttons disappear when viewing the site.

Could you share the serialized block to test it? Maybe you are connecting to a custom field that doesn't exist in the context of the homepage, and that's why the blocks are not rendering properly. Anyway, if that's an issue, I believe it should be related to the PHP logic.

> When in a template, for paragraph and heading fields connected to a field, let's not dim the cololur at all. It should be accurate to whatever styling is applied to the block

Makes sense. I was using the `placeholder` attribute, but I belive I can directly change the content. I'll give that a try.

> When in a template, for images, we could still show the alternative text field except it would be disabled and with the default value of the field inserted.

I wanted to explore that option as well. There is a challenge because the options in the sidebar don't appear until an image has been uploaded. But we need to figure out how to solve that.

> I don't see any design blockers for 6.5 here. During beta testing we might pick up the need to better highlight blocks that are connected to a field in some way.

I agree 🙂 I believe it would be nice to land a simpler UI first to have some testing but we can iterate over it during the upcoming weeks.

> There are e2e test failures logged for Pattern Overrides so it needs to be investigated. I suspect, it’s no longer possible to edit content for the blocks that have overrides enabled in the pattern.

Yep, I still have to take a deeper look at how it integrates with patterns because it might break things. Maybe we can check if the `source === 'pattern'` and skip any changes in that case.

SantosGuillamot committed:
Don't use placeholder if attribute is `src` or `href`

SantosGuillamot committed:
Always share placeholder in case meta is empty

SantosGuillamot committed:
Remove `keyToLabel` and use just label

SantosGuillamot committed:
Remove source component until it is needed

SantosGuillamot committed:
Use translations in the source label

SantosGuillamot commented:
> There are e2e test failures logged for Pattern Overrides, so it needs to be investigated. I suspect, it’s no longer possible to edit content for the blocks that have overrides enabled in the pattern.

Regarding this, we might need a way to know if we should lock the content, or not, depending on the source. My initial idea was to let the source decide that in the attributes. We could include a new parameter that we can later read. For example, in the case of a paragraph:

```
"metadata": {
    "bindings": {
      "content": {
        "source": {
          "name": "post_meta",
          "attributes": { "value": "text_custom_field", "lockEditorUI": true }
        }
      }
    }
  }
```

Would that make sense for this initial pull request until we have a better way to handle it?

SantosGuillamot committed:
Move `select` inside `useSource`

gziolo commented:
```js
metadata": {
    "bindings": {
      "content": {
        "source": {
          "name": "post_meta",
          "attributes": { "value": "text_custom_field" }
        },
        "lockEditorUI": true
      }
    }
  }
```

This is an intriguing idea. If we were to go with that approach then it should default to `true` so it only needs to get provided when the source wants to opt out like Pattern Overrides.

It would also justify using the `source` subkey for every connected attribute.

SantosGuillamot commented:
Okay, so if we go with the `lockEditorUI` property (or however we want to name it), it could work this way:

**Phase 1: read-only**
For this initial release, where users won't be able to edit the value of the custom fields, **we lock the editor controls by default** but other sources like pattern overrides can set it to false like this:

```
metadata": {
    "bindings": {
      "content": {
        "source": {
          "name": "post_meta",
          "attributes": { "value": "text_custom_field" }
        },
        "lockEditorUI": false
      }
    }
  }
```

**Phase 2: allow users to edit the value of the custom field**
In the future, we may want to allow users to bind the value of the attribute and the custom field and edit it directly from the UI. In that case, **the `lockEditorUI` property would become `false` by default**. And anyone can lock the UI if wanted setting it to `true`.

For other sources like pattern overrides that added the `lockEditorUI: false`, it will still work as expected.

---
Does it make sense? Any other approaches to this?

SantosGuillamot committed:
Read `lockEditorUI` prop and add it for patterns

michalczaplinski commented on packages/editor/src/bindings/post-meta.js:
```diff
@@ -0,0 +1,45 @@
+			placeholder: metaKey,
+			useValue: [ metaValue, updateMetaValue ],
+		};
+	},
```
Do we have to return an object with the `useValue` key? I see that the `useValue`
key is immediately destructured in
https://github.com/WordPress/gutenberg/blob/a4dc34a9eaa717c953b0393c82398ec28e9c62c7/packages/block-editor/src/hooks/use-bindings-attributes.js#L63-L64
and not used anywhere else.

Can we not just return `metaValue` and `updateMetaValue` directly?

I see that `updateMetaValue` is not currently used so we could even skip returning it for now, right?

michalczaplinski replied:
Additionally, I'm not sure if it's acceptable that `useSource()` returns a
different object depending on the code path. Right now, we return the
`{placeholder}` if it's a template and `{useValue}` otherwise.

Ideally, the return type should always be the same.

michalczaplinski replied:
> Additionally, I'm not sure if it's acceptable that useSource() returns a
> different object depending on the code path. Right now, we return the
> `{ placeholder }` if it's a template and `{ useValue }` otherwise.

> Ideally, the return type should always be the same.

Oh, nevermind. I see you've already updated it. 👍

SantosGuillamot replied:
> Can we not just return metaValue and updateMetaValue directly?

I used that trying to keep the "hooks syntax", but I'm fine returning just `metaValue` and `updataMetaValue` if that works and it is preferred.

michalczaplinski replied:
If the main motivation was to similarity to React hooks syntax then I would argue that it's similar but not the same so might be even more confusing. `useValue` here is not a function that returns an array but simply an array.

> returning just metaValue and updataMetaValue if that works

That would be my preference but I'm happy to hear other opinions 🙂

retrofox replied:
I would consider this option, however, detecting and propagating changes without using a React hook requires a different strategy, likely through subscribing to property value changes.

SantosGuillamot committed:
Move logic to lock editing directly to RichText

michalczaplinski commented on packages/block-editor/src/hooks/use-bindings-attributes.js:
```diff
@@ -0,0 +1,161 @@
+								}
+							}
+
+							if ( metaValue ) {
```
If the meta value is falsy, like `""` this will fail to use the meta value. So,
e.g. the paragraph block will appear with the original content but won't be
editable.

This is a somewhat confusing edge case because the there is no value that we can
show to the user in the editor to indicate that it's connected to a custom field.

SantosGuillamot replied:
In one of the refactors, assuming there is always a placeholder, I added [this logic](https://github.com/WordPress/gutenberg/blob/add/block-bindings-support-in-the-editor/packages/block-editor/src/hooks/use-bindings-attributes.js#L60-L76) to use it when the meta value is falsy. That way, we don't show either the original content or an empty attribute. Do you think it makes sense? Is it the issue you were referring to?

michalczaplinski replied:
yes, perfect 👍 . I've tested it quickly and it solves the issue.

SantosGuillamot commented:
> If we lock the editing UI for Post Meta by serializing lockEditorUI in the block attributes, how do you envision unlocking the UI for these blocks later when the UI is ready to use?

My initial idea was to change the conditionals we are adding. For example, if we take a look at the rick text one: [link](https://github.com/WordPress/gutenberg/blob/add/block-bindings-support-in-the-editor/packages/block-editor/src/components/rich-text/index.js#L303-L315).

Right now, the conditional is `settings.lockEditorUI !== false`. That means that if `lockEditor` is not defined (by default), it passes that conditional and set `shouldDisableEditing` to `true`.

In the future, we can change that to `settings.lockEditorUI === true`. In that case, if `lockEditor` is not defined, it won't pass the conditional and `shouldDisableEditing` will remain `false`.

As for this initial phase we have a controlled list of core blocks, I believe that shouldn't be an issue, right?

SantosGuillamot committed:
Improve `useSelect` destructuring

michalczaplinski reviewed (commented):
I think it's close. I only have a couple of comments. 

For me, only the last two are blocking:
- https://github.com/WordPress/gutenberg/pull/58085#discussion_r1463180899
- https://github.com/WordPress/gutenberg/pull/58085#discussion_r1463226796

SantosGuillamot committed:
Load all image controls if attributes are bound

SantosGuillamot committed:
Remove unnecessary condition

gziolo commented:
> In the future, we can change that to settings.lockEditorUI === true. In that case, if lockEditor is not defined, it won't pass the conditional and shouldDisableEditing will remain false.
>
> As for this initial phase we have a controlled list of core blocks, I believe that shouldn't be an issue, right?

Let's clarify it further as I'm not following the idea.

In version 1:
- Pattern Override will have to resolve `lockEditorUI` to `false.
- Post Meta will have to resolve `lockEditorUI` to `true.`

Since the value differs, what should be serialized in the HTML serialized to the database to represent these cases?
Would `lockEditorUI` ever differ between individual block bindings of the same type?

SantosGuillamot commented:
> Since the value differs, what should be serialized in the HTML serialized to the database to represent these cases?

Patterns will have to add the `lockEditorUI` property and set it to `false`, while `Post Meta` wouldn't need it because it's the default.

_Patterns_
```
<!-- wp:paragraph {"metadata":{"id":"ABCD","bindings":{"content":{"source":{"name":"pattern_attributes"},"lockEditorUI":false}}}} -->
```

_Post meta_
```
<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":{"name":"post_meta","attributes":{"value":"text_custom_field"}}}}}} -->
```

In the future, once we want to change the default to `false`, we just change the internals. And it would mean:
* Old patterns with the `lockEditorUI` property should keep working.
* Post Meta should work without `lockEditorUI`.
* New patterns don't need to include the `lockEditorUI` because that'd be the default then.

Anyway, this is just the way I thought about it. But we can implement it however we want based on feedback or use a totally different method.

> Would lockEditorUI ever differ between individual block bindings of the same type?

I believe it'd be good to be able to differ between block attributes if that's what you mean. For example, I may want to lock the text of the button but not the url.

SantosGuillamot committed:
Move `lockAttributesEditing` to source definition

SantosGuillamot commented:
I made [this commit](https://github.com/WordPress/gutenberg/pull/58085/commits/e6a5a4de52a2960e715a1cdf4e20170570d0062f#diff-cf4447b4134dbf24fa2819703999f0131449fb79a0611bd544c5e05fd80aa36bR41) to reflect another possibility. The idea is that by default, the attributes are editable unless the source specifies the opposite in the source registration.

In the case of Post Meta, in this initial phase, it can use `lockAttributesEditing: true` and remove that part once we want to integrate everything in the UI.

In the case of Pattern overrides, as they don't register anything for the editor because it is using its own solution, that property doesn't exist, so the blocks keep working as they do now.

Let me know if this alternative makes more sense.

ellatrix commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -289,6 +299,24 @@ export function RichTextWrapper(
 		anchorRef.current?.focus();
 	}
 
+	const bindings = getBlockAttributes( clientId )?.metadata?.bindings;
```
You should subscribe to these selector inside a `useSelect` callback. When the attributes change, the component will currently not re-render.

SantosGuillamot replied:
Do you mean something like this?

```js
const bindings = useSelect(
	( select ) => {
		return select( blockEditorStore ).getBlockAttributes( clientId )
			?.metadata?.bindings;
	},
	[ clientId ]
);
```

gziolo replied:
It also looks like there is some room for improvements to avoid computations in the case when there are no bindings set.

SantosGuillamot replied:
I've moved most of the logic to the initial `useSelect` callback that returns earlier if the block isn't selected: [link](https://github.com/WordPress/gutenberg/pull/58085/commits/7c1ca5a2bf0a1fa8a23bc56938ad2e6eddaa5c58). Let me know if that's what you had in mind.

> It also looks like there is some room for improvements to avoid computations in the case when there are no bindings set.

Regarding this, the only optimization I see after the last modification is to only get the `getBlockBindingsSource` function when bindings exist ([link](https://github.com/WordPress/gutenberg/blob/add/block-bindings-support-in-the-editor/packages/block-editor/src/components/rich-text/index.js#L136)) with something like this:

```
const { getBlockBindingsSource } = blockBindings
			? unlock( select( blockEditorStore ) )
			: {};
```

Would that make sense? Which other improvements do you see?

gziolo commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -289,6 +299,24 @@ export function RichTextWrapper(
+			// If any of the attributes with source "rich-text" is part of the bindings,
+			// has a source with `lockAttributesEditing`, disable it.
+			if (
+				blockTypeAttributes?.[ attribute ]?.source === 'rich-text' &&
```
This could also possibly lock RichText fields that aren't connected, if there is one that is connected and other that aren't. It isn't an issue for the blocks that have binding allowed, but I'm raising it as a future issue.

SantosGuillamot replied:
Are there any blocks with two attributes connected to "rich-text"? If so, is there a way to know which attribute the RichText corresponds to from the RichText?

gziolo replied:
Sure, Pullquote would be an example. It looks like the `identifier` props was used in the past whenever `RichText` needed to distinguish between wired attributes:

https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/pullquote/edit.js

@ellatrix, is it still supported? Have you ever considered moving the control for the block attribute instead of passing down the value to the `RichText` component living in the block editor? It would be convenient in this particular case.

ellatrix replied:
What do you mean with moving the control? Yes identifier is currently used almost everywhere, it's needed to store selection

gziolo replied:
Instead of

```js
<RichText
	tagName="pre"
	identifier="content"
	preserveWhiteSpace
	value={ content }
	onChange={ ( nextContent ) => {
		setAttributes( {
			content: nextContent,
		} );
	} }
	...   
```

we could have:

```js
<RichText
	tagName="pre"
	attributeName="content"
	preserveWhiteSpace
	...   
```

so `value` and `onChange` could be automatically handled internally (when not provided) based on the `attributeName` (we could keep the `identifier` as name).

ellatrix replied:
Yes, I considered that, but the ship has kind of sailed. We can try it again if you like, I don't remember the obstacles.

gziolo replied:
I noticed that @SantosGuillamot needs here to identify the attribute name that is being connected to put it in read-only mode and prevent doing it for other RichTexrt fields if they exist in the same block. Using `identifier` would help to refactor this code to avoid looping over attributes in the first place. At the same time, it triggered my thinking that since we need to override the value for the bound attribute, maybe we could skip passing it as `value` altogether.

@SantosGuillamot, the `value` and `onChange` is out of scope for this PR. I wanted mostly to confirm if we can use `identifier` here.

ellatrix replied:
You can use identifier, but it's not guaranteed to exist :)

gziolo replied:
We could always make it a requirement for block binding. At least for v1.

gziolo replied:
The discussion continues in https://github.com/WordPress/gutenberg/issues/58233.

ellatrix commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -376,7 +404,7 @@ export function RichTextWrapper(
 					anchorRef,
 				] ) }
-				contentEditable={ true }
+				contentEditable={ ! shouldDisableEditing }
```
A few lines down you'll need to set the tab index to:

```js
				tabIndex={
					props.tabIndex === 0 && ! shouldDisableEditing
						? null
						: props.tabIndex
				}
```

SantosGuillamot replied:
Thanks for the tip 🙂 Done in [this commit](https://github.com/WordPress/gutenberg/pull/58085/commits/42462609211628115cd9b03cb4758228a87ab103).

SantosGuillamot committed:
Move `useSelect` into existing hook

SantosGuillamot committed:
Fix `RichText` not being selected on click

SantosGuillamot committed:
Lock button and image controls only when selected

artemiomorales reviewed (commented):
While not a blocker for this PR, I started looking at accessibility via keyboard navigation and NVDA in Windows Firefox.

Some notes:

1. As far as I can see, it'd be good to announce to users that fields are noneditable in both _navigation mode_ and _edit mode_.

2. Additionally, we could consider having special handling for when users start typing into a paragraph expecting it to be editable, for example, without sending them into the block settings and making them lose context.

3. For disabled fields like image `alt`, since users may be expecting to see them as editable in the block settings like in normal blocks, we should consider announcing that those fields are noneditable.

4. While not really related to this PR, how to edit the content of the buttons just using the keyboard was unclear to me due the wrapping `buttons` block, and if it's indeed an issue, we can take that into account when determining any improvements we make (either improving it ourselves or coordinating with others who may have this on their radar).

The first two issues appear to be present with the pattern overrides as well; addressing these issues as a general effort around addressing improving accessibility in noneditable fields with screen readers and keyboard navigation might make sense.

Once this lands, we could open a separate issue to continue the discussion and gather more recommendations from accessibility folks.

Here's a video going over the points above. Note: I pause from speaking throughout so I don't talk over the screen reader.

Would be happy to hear additional thoughts or insights on any of this as well 🙏

https://github.com/WordPress/gutenberg/assets/5360536/ba2259db-fcb2-418e-8980-0362d59434ef

talldan commented:
> Phase 2: allow users to edit the value of the custom field
> In the future, we may want to allow users to bind the value of the attribute and the custom field and edit it directly from the UI. In that case, the lockEditorUI property would become false by default. And anyone can lock the UI if wanted setting it to true.

At that point it might be worth removing `lockEditorUI` completely. 🤔 

If that's the case, I guess the APIs could be kept private until that point, or there could be a separate private API to configure this.

The changes in this PR look ok to me, though I think they raise a lot of questions for the future about a more maintainable way to implement this kind of feature. FWIW, there are also similar concerns for patterns, where we're finding `contentOnly` to be too inflexible.

ellatrix commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -141,9 +150,17 @@ export function RichTextWrapper(
 			selectionEnd: isSelected ? selectionEnd.offset : undefined,
 			isSelected,
+			blockBindings,
+			getBlockBindingsSource,
```
For this it's better to use `const { getBlockBindingsSource } = useSelect(...)`

SantosGuillamot replied:
You mean outside the current `selector`, right? I can do that 🙂

SantosGuillamot replied:
I've changed that as part of [this commit](https://github.com/WordPress/gutenberg/pull/58085/commits/aaf8652d0e1facf6f1cd7431a4f96819e1f59c1b).

ellatrix commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -289,6 +306,23 @@ export function RichTextWrapper(
+				getBlockBindingsSource( args.source.name )
+					?.lockAttributesEditing
+			) {
+				shouldDisableEditing = true;
```
Could this potentially all be calculated in the useSelect callback? I'm not sure how much the bindings change, which would cause the component to re-render.

SantosGuillamot replied:
I can give that a try. In this initial version, bindings shouldn't change unless users modify them through the Code Editor. But I'll try to move it in the `useSelect` callback.

SantosGuillamot replied:
I've done it in [this commit](https://github.com/WordPress/gutenberg/pull/58085/commits/aaf8652d0e1facf6f1cd7431a4f96819e1f59c1b). Let me know if that's not what you had in mind.

ellatrix commented on packages/block-editor/src/hooks/use-bindings-attributes.js:
```diff
@@ -0,0 +1,148 @@
+	if ( ! ( settings.name in BLOCK_BINDINGS_ALLOWED_BLOCKS ) ) {
+		return settings;
+	}
+	settings.edit = createEditFunctionWithBindingsAttribute()( settings.edit );
```
What stops us from using the new "api" here? This will cause a bit of a performance regression.

SantosGuillamot replied:
Is it possible to modify the `attributes` and `setAttributes` of the `BlockEdit` component with the new API? If that's possible I believe I could change it.

ellatrix commented on packages/block-editor/src/hooks/use-bindings-attributes.js:
```diff
@@ -0,0 +1,148 @@
+		settings.usesContext = usesContextArray;
+		return settings;
+	}
+);
```
This filter is fine to leave as is.

ellatrix commented on packages/block-editor/src/store/private-selectors.js:
```diff
@@ -286,3 +286,11 @@ export const hasAllowedPatterns = createSelector(
+}
+
+export function getBlockBindingsSource( state, sourceName ) {
+	return state?.blockBindingsSources?.[ sourceName ];
```
Optional chaining seems unnecessary here

SantosGuillamot replied:
Done in [this commit](https://github.com/WordPress/gutenberg/pull/58085/commits/54c313dc3366bd263c8d67b517813c3e368990f8).

ellatrix reviewed (commented):
Generally, it feels a bit strange to have to change each block to account for the block bindings. But I guess you've considered all that.

gziolo reviewed (commented):
E2e tests are fixed and the most important feedback I left was addressed. I mostly want to remove "request changes" review status as I want to have time today to review the PR.

SantosGuillamot commented:
> Generally, it feels a bit strange to have to change each block to account for the block bindings. But I guess you've considered all that.

The end goal is to enable bindings for any block, so all of them should be able to understand the bindings. We did it that way because we assume it was needed, but it'd be great to know other alternatives that might be better.

SantosGuillamot committed:
Remove unnecesarry optional chaining

SantosGuillamot committed:
Move `shouldDisableEditing` logic inside callback

SantosGuillamot committed:
Merge branch 'trunk' into add/block-bindings-support-in-the-editor

SantosGuillamot committed:
Fix formatting issue

SantosGuillamot commented:
I think I’ve addressed most of the feedback, especially the blockers. There are some missing decisions/improvements, but I believe most of them can be done in other pull requests. These are the things I identified:

- Accessibility improvements: [link](https://github.com/WordPress/gutenberg/pull/58085#pullrequestreview-1840030495). Artemio made great points about things that can be improved regarding accessibility. It seems it is not only related to this pull request, but to other parts of the editor as well. For example, I tested the Post Date block, which has a similar behavior. I can work on a new PR and reach the accessibility team to see how we should manage it.
- Can we use use the new block-editor hooks API instead of changing `settings.edit`: [link](https://github.com/WordPress/gutenberg/pull/58085#discussion_r1464611692)? If possible, we may change the filters used. Anyway, I thing that can be done later as it only affects the internal code and (maybe) the performance.
- Improve the way we are handling rich-text attributes to make it work with blocks that accept multiple rich texts: [link](https://github.com/WordPress/gutenberg/pull/58085#discussion_r1463432529). Right now it isn’t a blocker because the list of blocks that can use bindings is restricted.
- Should we have `metaValue` and `setMetaValue` instead of `useSource: [ metaValue, setMetaValue ]`: [link](https://github.com/WordPress/gutenberg/pull/58085#discussion_r1463150661)? Whatever we decide can be changed later because the API will remain private.
- Should we add some UI to remove the bindings in this initial version:  [link](https://github.com/WordPress/gutenberg/pull/58085#discussion_r1462805970)? In my opinion, this can be added as an improvement later if we consider it necessary.
- Should the placeholders show the value of the key or should they show something different: [link](https://github.com/WordPress/gutenberg/pull/58085#discussion_r1462809261)? Again, I believe this is a small change that can be done in other pull request.

Would you consider any of those a blocker? Is there anything else I’m missing?

michalczaplinski commented:
I see an issue: The content of the block is not updated when you edit the metadata in the block. You have to make some _other_ edits in order for the content to be updated **on the frontend**. In the editor, the value is updated correctly.


https://github.com/WordPress/gutenberg/assets/5417266/0581e973-edbf-4b94-895b-65f8ce3546ae

michalczaplinski commented:
> I see an issue: The content of the block is not updated when you edit the metadata in the block. You have to make some other edits in order for the content to be updated on the frontend. In the editor, the value is updated correctly.

I can reproduce the same issue with the Image block.

SantosGuillamot commented:
I've been trying to reproduce it following the exact same process, but I haven't been lucky. I'll keep trying to be able to identify the issue better.

artemiomorales commented:
> I see an issue: The content of the block is not updated when you edit the metadata in the block. You have to make some other edits in order for the content to be updated on the frontend. In the editor, the value is updated correctly.

Hmm, am unable to reproduce this one following the same steps 🤔

artemiomorales reviewed (approved):
I'm approving this PR on the grounds that [removing the experimental flag](https://github.com/WordPress/gutenberg/pull/58089#event-11587042002) has already been merged and the premise of this PR, which is to disable editing of bound fields and improve the UX, is working as intended as far as I can see.

I think we can continue addressing the improvements Mario outlined as well as fix any outstanding issues as feedback comes in.

michalczaplinski commented on packages/block-editor/src/hooks/use-bindings-attributes.js:
```diff
@@ -0,0 +1,148 @@
+	'core/paragraph': [ 'content' ],
+	'core/heading': [ 'content' ],
+	'core/image': [ 'url', 'title', 'alt' ],
+	'core/button': [ 'url', 'text' ],
```
Should we also add the `linkTarget` here?

gziolo replied:
There were some issues with `linkTarget`. Let's keep it out of the scope for now.

michalczaplinski reviewed (approved):
It looks good to me, awesome work, Mario! 👍 

We can address any issues in follow-up PRs.

michalczaplinski commented:
> I see an issue: The content of the block is not updated when you edit the metadata in the block. You have to make some other edits in order for the content to be updated on the frontend. In the editor, the value is updated correctly.

This was caused by the recent changes in `trunk` where the GB experiment was removed (https://github.com/WordPress/gutenberg/pull/58089). This PR was branched off before that, so it required the experiment to be enabled. 

In summary, the issue should not be present in `trunk` anymore after the merge.

ellatrix commented on packages/block-editor/src/hooks/use-bindings-attributes.js:
```diff
@@ -0,0 +1,148 @@
+							const {
+								placeholder,
+								useValue: [ metaValue = null ] = [],
+							} = source.useSource(
```
Is this a hook? You cannot use hooks inside conditions and loops

SantosGuillamot replied:
In the case of post meta source, it uses `useSelect` and `useEntityProp`: [link](https://github.com/WordPress/gutenberg/pull/58085/files#diff-cf4447b4134dbf24fa2819703999f0131449fb79a0611bd544c5e05fd80aa36bR15-R39). Although each source could do whatever they want.

If that's a problem, do you have any ideas on how that should be handled? We need to iterate through the different bindings and get the value from each of the sources.

ellatrix replied:
The only solution is to probably mount sub components, get the result, and set some state in this component, similar to how it is done here: https://github.com/WordPress/gutenberg/blob/fcf0b5d4cb15ec521c562fc367c3fa6c1220013c/packages/block-editor/src/hooks/utils.js#L537

But I'd love to hear other people's thoughts

michalczaplinski replied:
Good catch @ellatrix! Yup, we'll have to change it because it is a hook.

retrofox commented on packages/block-editor/src/hooks/use-bindings-attributes.js:
```diff
@@ -0,0 +1,148 @@
+						);
+
+						if ( source ) {
+							// Second argument (`updateMetaValue`) will be used to update the value in the future.
```
Any idea or advance of how it's going to implement the `updateMetaValue` usage?

SantosGuillamot replied:
The technical implementation is almost there. However, there are many things to take into account, and it was decided not to rush things for the upcoming 6.5 to ensure that whatever we land is stable enough.

retrofox commented on packages/block-editor/src/hooks/use-bindings-attributes.js:
```diff
@@ -0,0 +1,148 @@
+							// Second argument (`updateMetaValue`) will be used to update the value in the future.
+							const {
+								placeholder,
+								useValue: [ metaValue = null ] = [],
```
Is it `metaValue` a proper name here? In theory, the source can be arbitrary meaning it could belong to metadata, but also to very different origins. 
Could we consider using something like `sourceValue`, and `setSourceValue`?

gziolo replied:
Good feedback. On the server, there is `get_value_callback` registered for every block bindings source.

Mamaduka commented on packages/block-editor/src/hooks/use-bindings-attributes.js:
```diff
@@ -0,0 +1,148 @@
+					<BlockEdit
+						key="edit"
+						attributes={ updatedAttributes }
+						setAttributes={ ( newAttributes, blockId ) =>
+							registry.batch( () =>
+								updateBlockAttributes( blockId, newAttributes )
+							)
+						}
```
* The blocks expect `setAttributes` to have a stable reference between rerenders. This override breaks that.
* What's the reason for batching the single action? Technically, it shouldn't make a difference.


P.S. There is no need to wrap the returned `BlockEdit` in the fragment.

gziolo replied:
@Mamaduka, I think the issue at the moment is that users will only change the attribute when it is not bound. When the value is sourced from the block binding then the attribute becomes readonly. In effect, the batching isn't really necessary and even there should be no override set until it's possible to edit the external source in UI.

By the way, feel free to refactor the code here if you have some ideas how to optimize it.

Mamaduka replied:
> I think the issue at the moment is that users will only change the attribute when it is not bound. When the value is sourced from the block binding then the attribute becomes readonly.

Is that handled somewhere else? Because I don't see that logic here. If the block calls the `setAttribute` override, it will update the attribute, bound or not.

Sorry, I'm not super familiar with the internals of Block Binding API. I just came across this code while looking for something else.

gziolo replied:
> it will update the attribute, bound or not

There is no way to trigger the update from UI at the moment for supported blocks and their selected attributes. Direct changes to the store don't matter because the value will get replaced on the frontend anyway. 

You are totally right that batching is not needed at the moment and most likely the override for `setAttributes` is premature.

Mamaduka replied:
Thanks for confirming. I can push refactoring PR later today.
